### DISCUSSION TITLE
Suggestion: add/document a 'using' function for transactions

### DISCUSSION THREAD

**Discussion by danwkennedy** [arangodb/arangojs#786] [Ideas]
Found myself using this approach and thought it'd be useful for others. I'm not sure where it'd fit in the arangojs API's but it's still pretty handy.

To help developers using the streaming transaction API's, we can write a pretty simple "using" function like they have in C#:

```js
module.exports = async function usingTransaction(transaction, action) {
  try {
    // run action passing the transaction through
    const ret = await action(transaction);

    // if the action returned, commit and return the result if any
    await transaction.commit();
    return ret;
  } catch (err) {
    // if the action failed, abort and throw
    await transaction.abort();
    throw err;
  }
};
```

This would ensure that we don't forget to actually commit/abort the transactions. An example use case for this is:

```js
async function transactionExample(db) {
  const collection1 = db.collection('collection1');
  const collection2 = db.collection('collection2');

  const trx = await db.beginTransaction([collection1, collection2]);
  
  return usingTransaction(trx, async (transaction) => {
    await transaction.run(() => collection1.save({});
    await transaction.run(() => collection2.save({});
  });
}
```

**Comment by pluma** [comment]
I'm intrigued but I'm not sure I like the API as described. I'm not sure if this would work better as a standalone helper function or a method on the Transaction or Database class.

**Comment by danwkennedy** [comment]
Yeah, it could probably benefit from being an internal function. It could simplify the example above to something like this:

```js
async function transactionExample(db) {
  const collection1 = db.collection('collection1');
  const collection2 = db.collection('collection2');
  
  await db.transaction([collection1, collection2], async trx => {
    await transaction.run(() => collection1.save({});
    await transaction.run(() => collection2.save({});
  });
}
```

**Comment by danwkennedy** [comment]
A potentially more elegant solution would also be to use an async generator where you `yield` db operations that you want part of the transaction and the runner function calls it inside of a `transaction.run()` call and pass the results through.

I haven't tested this but this would be potentially how it would look:

```js

async function transactionExample(db) {
  const collection1 = db.collection('collection1');
  const collection2 = db.collection('collection2');
  
  await db.transaction([collection1, collection2], async trx => {
    yield () => collection1.save({});
    yield () => collection2.save({});
  });
}

```
